### ISSUE TITLE
Intermittent health check failure causes unreachable downstream clusters

### ISSUE THREAD

**Issue by SheilaghM** [rancher/rancher#34819]
SURE-3343, SURE-3344

Rancher Cluster:
Rancher version:2.5.9
Number of nodes:1500
Node OS version: RancherOS- v1.5.6
RKE/RKE2/K3S version: RKE
Kubernetes version: 1.18.6

Downstream Cluster:
Number of Downstream clusters: 150
Node OS: RancherOS-v1.5.6
RKE/RKE2/K3S version: RKE
Kubernetes version:v1.18.20
CNI:
 
Other:
Underlying Infrastructure: Azure
Any 3rd party software installed on the nodes: NA
Customer’s main time zone: UTC +2

**Describe the bug**
System fails intermittently with message:

Cluster health check failed: Failed to communicate with API server: Get "https://[api-server]/api/v1/namespaces/kube-system?timeout=45s": context deadline exceeded

The clusters recover after a few seconds and the errors occur at random intervals. 

**To Reproduce**
We currently don't have repro steps but the issue might be happening while applying a big ConfigMap

**Result**
Random downstream clusters are intermittently unreachable for 5~10 min and auto recovering after 5~10 min.

**Expected Result**
The cluster should not enter the unreachable state.

**Comment by junkiebev** [comment]
Having the same issue. Upgrading monitoring to .22￼ on all downstream clusters and projects seemed to help, but it's still happening.

**Comment by ibuildthecloud** [comment]
While investigating this issue it has been found that if the connection buffer fills up it is possible to block all goroutines related to one cluster.  Under the right conditions this can cause this health check failure issue.  It's hard to reproduce the intermittent health check failures but the underlying buffer issue can be reproduce fairly easily.

Setup Rancher 2.6.3 managing a single cluster.  Ensure that the networking connection between Rancher and the managed cluster is fast/low latency.  For example just run both in the same cloud.

Then create a deployment in the downstream cluster that has a 100MB file in the container.

Now from a client run `kubectl cp` in parallel using a kubeconfig file from Rancher so that the traffic is going through Rancher's tunnels.  If you run about 10 in parallel you should see they will sometimes fail and all block/timeout before the full 100mb is copied.  For this to work it is important that the connection from client to Rancher be slower than the connection from Rancher to managed cluster. I tested this by running the client from my laptop.

I used the below simple script to run `kubectl cp` in parallel.

```bash
#!/bin/bash

for i in $(seq 10); do
    (
        time kubectl cp ${POD}:/test test.$i
        ls -l test.$i
        echo done $i
    ) 2>&1 | tee test.$i.log &
done
```

**Comment by slickwarren** [comment]
@ibuildthecloud 
I’ve followed the instructions and have `kubectl cp` running in parallel. The loop seems to finish, but none of the actual images seem to have copied fully. Total file size is `354M`, but the parallel process always downloads less than 100 each when ran in parallel. So that’s the issue, right? Just wanted to make sure that there’s no explicit error thrown or anything.
from the screenshot, `bionic-#.img` is from parallel runs, and `bionic-from-pod.img` is from a single `kubectl cp`

<img width="553" alt="Screen Shot 2022-01-07 at 2 21 14 PM" src="https://user-images.githubusercontent.com/16691014/148617146-f61b2892-9b09-41be-8636-5a44b7958b5f.png">

**Comment by ibuildthecloud** [comment]
Correct. You won’t see an error except that the copies will stop prematurely. kubectl will print something about an unexpected EOF.

**Comment by slickwarren** [comment]
**I was not able to reproduce this issue**


hmm. Actually, I wasn't able to reproduce this issue. After waiting a little longer, it looks like all of them finished. I tried again with 25 in parallel but those also finished fine and the file sizes match the original. All finished with a status like 
```
kubectl cp  bionic-$i.img  2.89s user 3.64s system 1% cpu 10:39.61 total
-rw-r--r--  1 calebwarren  staff  371523584 Jan  7 15:50 bionic-24.img
done 24

[11]    52112 done       ( time kubectl cp  bionic-$i.img; ls -G -l bionic-$i.img; echo done $i; ) 2>& | 
        52113 done       tee test.$i.log
```
none of them have any EOF or errors that I see. 


Here's what I did to try and reproduce:
* deployed rancher 2.6.3 in us-west-1
* deployed a downstream cluster in us-west-1, single node
  * deployed workload
  * added a large file (>100MB) to the pod
* download kubeconfig for the downstream cluster to local machine, and ran the following:
```
for i in {15..40}; do\n    (
        time kubectl cp <rancher_pod>:home/bionic-server-cloudimg-amd64-cp1.img bionic-$i.img 
        ls -l bionic-$i.img
        echo done $i
) 2>&1 | tee test.$i.log &
done
```
* verified that ping  from client machine (laptop) to downstream (~15ms) was longer than rancher machine -> downstream (~1.5ms)

**Comment by davidnuzik** [comment]
<!-- ALWAYS be cautious to not include any sensitive info such as external IPs, FQDNs, passwords, keys, etc in either screenshots or any text, and to follow the QA Best Practices -->
 
**I believe I have successfully reproduced when running Rancher and downstream k3s on the same box. Please see details and Results section below.**
 
**Reproduction Environment:**
<!-- Reproduction is required if the issue was NOT filed by yourself otherwise it has to be reproduced (However there may be exceptions which will be communicated during testing) -->
Rancher version: v2.6.3
Rancher cluster type: single-node docker install
Docker version: 20.10
 
Downstream cluster type: k3s
Downstream K8s version: v1.22.5+k3s1 - running on the same box as the single-node rancher server
 
**Reproduction steps:**
<!-- Provide explanations for anything that may be confusing, exact CLI commands (if used), output of commands run (if any), relevant logs snippets, error messages, and screenshots if needed -->
Note: These steps are derrived from [Darren's comment here](https://github.com/rancher/rancher/issues/34819#issuecomment-1007600901).

0. Create a DigitalOcean VM. I did this with `ubuntu-20-04-x64` image in sf3 region with size `s-2vcpu-4gb`
`doctl compute droplet create dave-34819-4gb --tag-names dave-daily --image ubuntu-20-04-x64 --region sfo3 --size s-2vcpu-4gb --ssh-keys FINGERPRINT-HERE`
1. SSH into the box, install docker 20.10 and deploy single-node rancher 2.6.3 with docker.
`docker run -v ~/state:/var/lib/rancher -d -p 80:80 -p 443:443 --name two-six-three --restart=no --privileged rancher/rancher:v2.6.3` (note -v flag and --name are optional here; this is just how I operate...)
2. Install K3S (such as like `curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--write-kubeconfig-mode 0644 --node-external-ip EXTERNAL-IP-HERE" sh -`) (setting node-external-ip is probably not needed for this issue, just a habit)
Note, at this time the digitalOcean vm has almost 1GB free/cached memory which should be fine. (Swappiness is disabled OOTB).
3. Delete the traefik service (so that we can still access rancher frontend without any port conflicts) `kubectl delete svc traefik -n kube-system` (I might have been able to tell k3s at install-time to not deploy traefik; not sure but this works)
4. Login to the rancher frontend. Import this k3s cluster.
5. Create a deployment in the downstream k3s cluster. In my case I just deployed `ranchertest/mytestcontainer:latest`.
6. Exec shell into the container. Create a 100mb file on disk inside the container. I just did this at /
`fallocate -l $((100*1024*1024)) test` and then confirmed the file is as I expect it to be: `ls -lh *.txt` and confirmed it's indeed 100MB on disk:
```-rw-r--r-- 1 root root 100M Jan 12 19:17 test```
7. Now, on a client outside the LAN (which was my home computer, 1gbps down wan connection, Ethernet) get the downstream k3s cluster kubeconfig file and set it as kubeconfig for kubectl. Then, create a script on my local machine like as follows. This will ensure we run `kubectl cp` sort of in parallel (I think it's sort of because this has to loop, but it's almost like parallel because it takes time to download each); in theory this should reproduce the issue especially if the connection is slow enough.
```
#!/bin/bash

for i in $(seq 10); do
    (
        time kubectl cp ${POD}:/test test.$i
        ls -l test.$i
        echo done $i
    ) 2>&1 | tee test.$i.log &
done
```

**Results:**
So, I believe I have reproduced? @ibuildthecloud 

When I run the script I see that shell output starts with the following (which I did not expect):

```
 tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names
```

However, I then see unexpected EOF errors for all instances (I reproduced then?):

```
david@neo:~/kc-cp-test> error: unexpected EOF

real    0m24.732s
user    0m0.453s
sys     0m0.319s
-rw-r--r-- 1 david users 97965568 Jan 12 12:52 test.6
done 6
error: unexpected EOF

real    0m26.117s
user    0m0.350s
sys     0m0.396s
-rw-r--r-- 1 david users 99104256 Jan 12 12:52 test.2
done 2
error: unexpected EOF

real    0m28.817s
user    0m0.396s
sys     0m0.366s
-rw-r--r-- 1 david users 95718912 Jan 12 12:52 test.5
done 5

real    0m31.193s
user    0m0.417s
sys     0m0.378s
-rw-r--r-- 1 david users 104857600 Jan 12 12:52 test.3
done 3

real    0m31.742s
user    0m0.465s
sys     0m0.363s
-rw-r--r-- 1 david users 104857600 Jan 12 12:52 test.9
done 9
error: unexpected EOF

real    0m32.589s
user    0m0.429s
sys     0m0.336s
-rw-r--r-- 1 david users 101301760 Jan 12 12:52 test.7
done 7
error: unexpected EOF

real    0m32.737s
user    0m0.417s
sys     0m0.335s
-rw-r--r-- 1 david users 97205760 Jan 12 12:52 test.4
done 4
error: unexpected EOF

real    0m34.521s
user    0m0.405s
sys     0m0.396s
-rw-r--r-- 1 david users 103837184 Jan 12 12:52 test.10
done 10
error: unexpected EOF

real    0m34.549s
user    0m0.365s
sys     0m0.419s
-rw-r--r-- 1 david users 104461824 Jan 12 12:52 test.8
done 8
error: unexpected EOF

real    0m35.539s
user    0m0.453s
sys     0m0.357s
-rw-r--r-- 1 david users 104173056 Jan 12 12:52 test.1
done 1

```

And finally when I check file sizes on disk, some of the files are note full 100MB:

```
david@neo:~/kc-cp-test> ls -laht | grep -iv "log"
total 967M
-rw-r--r-- 1 david users 100M Jan 12 12:52 test.1
-rw-r--r-- 1 david users 100M Jan 12 12:52 test.8
-rw-r--r-- 1 david users 100M Jan 12 12:52 test.10
-rw-r--r-- 1 david users  93M Jan 12 12:52 test.4
-rw-r--r-- 1 david users  97M Jan 12 12:52 test.7
-rw-r--r-- 1 david users 100M Jan 12 12:52 test.9
-rw-r--r-- 1 david users 100M Jan 12 12:52 test.3
-rw-r--r-- 1 david users  92M Jan 12 12:52 test.5
-rw-r--r-- 1 david users  95M Jan 12 12:52 test.2
-rw-r--r-- 1 david users  94M Jan 12 12:52 test.6
drwxr-xr-x 1 david users  354 Jan 12 12:51 .
-rwxr-xr-x 1 david users  217 Jan 12 12:51 kc-cp-script.sh
drwxr-xr-x 1 david users 1.3K Jan 12 12:39 ..
```

**Additional Info:**
<!-- You can provide additional info here such explanations, details, or full logs, if needed -->
While running the script I see free memory on the DO host decline, but it doesn't appear to be detrimental. Typically always at least 400MB free/cached. Likely not an issue.

**Comment by davidnuzik** [comment]
<!-- ALWAYS be cautious to not include any sensitive info such as external IPs, FQDNs, passwords, keys, etc in either screenshots or any text, and to follow the QA Best Practices -->
 
**My validation checks Failed**
 
**Reproduction Steps:**
<!-- Reproduction is not required if the issue was filed by yourself else it has to be reproduced (of course there will be exceptions but they will be called out during a release testing) -->
I already reproduced this recently in the [prior comment above](https://github.com/rancher/rancher/issues/34819#issuecomment-1011404585). Also, when I reproduced I tested several times and had the same result each time (encountering EOF errors).

-----
 
**Validation Environment:**
<!-- If the Rancher version is from *-head then include the commitId from rancher server logs or from the UI - bottom left side -->
Rancher version: v2.5-head (effectively equivalent to v2.5.12-rc3 - rc3 was broken so I had to use v2.5-head; it should make no difference as both have the change)
Rancher cluster type: single-node docker install
Docker version: 20.10
 
Downstream cluster type: k3s
Downstream K8s version: v1.20.14+k3s1 (Rancher 2.5 should work with up to 1.20.x Kubernetes so this version was used for validation)
 
**Validation steps:**
<!-- Provide explanations for anything that may be confusing, exact CLI commands (if used), output of commands run (if any), relevant logs snippets, error messages, and screenshots if needed -->
Note: These steps are derrived from [Darren's comment here](https://github.com/rancher/rancher/issues/34819#issuecomment-1007600901).

0. Create a DigitalOcean VM. I did this with `ubuntu-20-04-x64` image in sf3 region with size `s-2vcpu-4gb`
`doctl compute droplet create dave-34819-4gb --tag-names dave-daily --image ubuntu-20-04-x64 --region sfo3 --size s-2vcpu-4gb --ssh-keys FINGERPRINT-HERE`
1. SSH into the box, install docker 20.10 and deploy single-node rancher 2.6.3 with docker.
`docker run -v ~/state:/var/lib/rancher -d -p 80:80 -p 443:443 --name two-five-head --restart=no --privileged rancher/rancher:v2.5-head` (note -v flag and --name are optional here; this is just how I operate...)
2. Install K3S (such as like `curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=1.20.14+k3s1 INSTALL_K3S_EXEC="--write-kubeconfig-mode 0644 --node-external-ip EXTERNAL-IP-HERE" sh -`) (setting node-external-ip is probably not needed for this issue, just a habit)
Note, at this time the digitalOcean vm has almost 1GB free/cached memory which should be fine. (Swappiness is disabled OOTB).
3. Delete the traefik service (so that we can still access rancher frontend without any port conflicts) `kubectl delete svc traefik -n kube-system` (I might have been able to tell k3s at install-time to not deploy traefik; not sure but this works)
4. Login to the rancher frontend. Import this k3s cluster.
5. Create a deployment in the downstream k3s cluster. In my case I just deployed `ranchertest/mytestcontainer:latest`.
6. Exec shell into the container. Create a 100mb file on disk inside the container. I just did this at /
`fallocate -l $((100*1024*1024)) test` and then confirmed the file is as I expect it to be: `ls -lh test` and confirmed it's indeed 100MB on disk:
```-rw-r--r-- 1 root root 100M Jan 14 17:55 test```
7. Now, on a client outside the LAN (which was my home computer, 1gbps down wan connection, Ethernet) get the downstream k3s cluster kubeconfig file and set it as kubeconfig for kubectl. Query for the name of the pod (`kubectl get pods -A`) and export that name into a var named `POD` like `export POD=eoftest-6bbd465565-lcxjx` for example. Then, create a script on my local machine like as follows and execute it. This will ensure we run `kubectl cp` sort of in parallel (I think it's sort of because this has to loop, but it's almost like parallel because it takes time to download each); in theory this should reproduce the issue especially if the connection is slow enough. When I tested I saw less EOF errors, but still encountered them.
```
#!/bin/bash

for i in $(seq 10); do
    (
        time kubectl cp ${POD}:/test test.$i
        ls -l test.$i
        echo done $i
    ) 2>&1 | tee test.$i.log &
done
```

**Results:**
I still encounter the EOF errors but I seem to encounter less of them. This is either due to the changes partially resolving the issue or perhaps the method in which I am testing could be the culprit. However, I am testing with the exact same way as I did when I initially reproduced the issue in my prior comment.

I still encounter this unexpected output, but then the script seems to work as expected. And it shows less EOF errors than previously when testing with v2.6.3.
```
 tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names                                                                                                                                                                                                                                   
tar: Removing leading `/' from member names
```

I then see some EOF errors. These are encountered on loop iteration 4,5,7, and 10.

```
real    0m40.880s
user    0m0.449s
sys     0m0.384s
-rw-r--r-- 1 david users 104857600 Jan 14 11:03 test.8
done 8
error: unexpected EOF

real    0m42.496s
user    0m0.413s
sys     0m0.403s
-rw-r--r-- 1 david users 103628288 Jan 14 11:03 test.5
done 5
error: unexpected EOF

real    0m42.640s
user    0m0.452s
sys     0m0.309s
-rw-r--r-- 1 david users 97392128 Jan 14 11:03 test.4
done 4
error: unexpected EOF

real    0m43.139s
user    0m0.384s
sys     0m0.380s
-rw-r--r-- 1 david users 97558016 Jan 14 11:03 test.7
done 7
error: unexpected EOF

real    0m43.199s
user    0m0.358s
sys     0m0.430s
-rw-r--r-- 1 david users 97320448 Jan 14 11:03 test.10
done 10

real    0m43.245s
user    0m0.463s
sys     0m0.376s
-rw-r--r-- 1 david users 104857600 Jan 14 11:03 test.1
done 1

real    0m43.676s
user    0m0.477s
sys     0m0.373s
-rw-r--r-- 1 david users 104857600 Jan 14 11:03 test.6
done 6

real    0m44.347s
user    0m0.425s
sys     0m0.409s
-rw-r--r-- 1 david users 104857600 Jan 14 11:03 test.3
done 3

real    0m44.402s
user    0m0.502s
sys     0m0.344s
-rw-r--r-- 1 david users 104857600 Jan 14 11:03 test.9
done 9

real    0m45.016s
user    0m0.428s
sys     0m0.398s
-rw-r--r-- 1 david users 104857600 Jan 14 11:03 test.2
done 2
```

And when I check sizes on disk I observe the following, as expected, due to the EOF errors:

```
david@neo:~/kc-cp-test> ls -laht | grep -iv "log"
total 978M
-rw-r--r-- 1 david users 100M Jan 14 11:03 test.2
-rw-r--r-- 1 david users 100M Jan 14 11:03 test.9
-rw-r--r-- 1 david users 100M Jan 14 11:03 test.3
-rw-r--r-- 1 david users 100M Jan 14 11:03 test.6
-rw-r--r-- 1 david users 100M Jan 14 11:03 test.1
-rw-r--r-- 1 david users  93M Jan 14 11:03 test.10
-rw-r--r-- 1 david users  94M Jan 14 11:03 test.7
-rw-r--r-- 1 david users  99M Jan 14 11:03 test.5
-rw-r--r-- 1 david users  93M Jan 14 11:03 test.4
-rw-r--r-- 1 david users 100M Jan 14 11:03 test.8
drwxr-xr-x 1 david users  392 Jan 14 11:03 .
drwxr-xr-x 1 david users  324 Jan 14 10:58 v2.6.3-test-results
drwxr-xr-x 1 david users 1.3K Jan 13 09:07 ..
-rwxr-xr-x 1 david users  217 Jan 12 12:51 kc-cp-script.sh
```

**Additional Info:**
<!-- You can provide additional info here such explanations, details, or full logs, if needed -->
While running the script I see free memory on the DO host decline again, but it doesn't appear to be detrimental. Typically always at least 400MB free/cached. Likely not an issue.

Rancher server logs snippet below. Notice this is similar to my snippet prior when I reproduced. This is probably because the snippet prior wasn't related to the actions being performed by the script. We now see remotedialer buffer exceeded errors. There are a ton of these errors, this is just a snippet of them.
```
2022/01/14 18:03:47 [ERROR] remotedialer buffer exceeded, length: 4215507
2022/01/14 18:03:47 [ERROR] remotedialer buffer exceeded, length: 4248275
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4206503
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4239271
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4272039
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4304807
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4337575
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4370343
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4403111
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4435879
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4468647
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4501415
2022/01/14 18:03:56 [ERROR] remotedialer buffer exceeded, length: 4509027
```

k3s server logs snippet below. I don't thinkany of this is relevant except maybe the last two lines.

```
Jan 14 18:03:17 dave-34819-4gb k3s[788250]: I0114 18:03:17.698615  788250 trace.go:205] Trace[1605878027]: "GuaranteedUpdate etcd3" type:*core.ConfigMap (14-Jan-2022 18:03:17.060) (total time: 638ms):
Jan 14 18:03:17 dave-34819-4gb k3s[788250]: Trace[1605878027]: ---"Transaction committed" 637ms (18:03:00.698)
Jan 14 18:03:17 dave-34819-4gb k3s[788250]: Trace[1605878027]: [638.370347ms] [638.370347ms] END
Jan 14 18:03:17 dave-34819-4gb k3s[788250]: I0114 18:03:17.698914  788250 trace.go:205] Trace[1447201942]: "Update" url:/api/v1/namespaces/fleet-system/configmaps/fleet-agent-lock,user-agent:fleetagent/v0.0.0 (linux/amd64) kubernetes/$Format,client:10.42.0.11 (14-Jan-2022 18:03:17.059) (total time: 638ms):
Jan 14 18:03:17 dave-34819-4gb k3s[788250]: Trace[1447201942]: ---"Object stored in database" 638ms (18:03:00.698)
Jan 14 18:03:17 dave-34819-4gb k3s[788250]: Trace[1447201942]: [638.932235ms] [638.932235ms] END
Jan 14 18:03:56 dave-34819-4gb k3s[788250]: E0114 18:03:56.170954  788250 upgradeaware.go:373] Error proxying data from client to backend: EOF
Jan 14 18:03:56 dave-34819-4gb k3s[788250]: E0114 18:03:56.175627  788250 upgradeaware.go:387] Error proxying data from backend to client: tls: use of closed connection
```

**Comment by sowmyav27** [comment]
PR: https://github.com/rancher/rancher/pull/36206

**Comment by sowmyav27** [comment]
Based on offline discussion with @davidnuzik @ibuildthecloud @deniseschannon closing this issue for 2.5.12 release. 
Note: We have not be able to full repro/validate this issue. We will have user test it using 2.5.12 and 2.6.3-patch1 and if there are any further fixes needed, they will be addressed separately.
